File: atopmall_web/goods_web/api/goods/goods.go
package goods

import (
	"context"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"

	"atopmall_web/goods_web/api"
	"atopmall_web/goods_web/forms"
	"atopmall_web/goods_web/global"
	"atopmall_web/goods_web/proto"
)

func GetGoodsList(ctx *gin.Context) {
	//查询商品列表（各种过滤）
	req := &proto.GoodsFilterRequest{}
	priceMinInt, _ := strconv.Atoi(ctx.DefaultQuery("pmin", "0"))
	req.PriceMin = int32(priceMinInt)

	priceMaxInt, _ := strconv.Atoi(ctx.DefaultQuery("pmax", "0"))
	req.PriceMax = int32(priceMaxInt)

	isHot := ctx.DefaultQuery("ishot", "0")
	if isHot == "1" {
		req.IsHot = true
	}
	isNew := ctx.DefaultQuery("isnew", "0")
	if isNew == "1" {
		req.IsNew = true
	}
	isTab := ctx.DefaultQuery("istab", "0")
	if isTab == "1" {
		req.IsTab = true
	}
	categoryId := ctx.DefaultQuery("c", "0")
	categoryIdInt, _ := strconv.Atoi(categoryId)
	req.TopCategory = int32(categoryIdInt)

	pages := ctx.DefaultQuery("p", "0")
	pagesInt, _ := strconv.Atoi(pages)
	req.Pages = int32(pagesInt)

	perNums := ctx.DefaultQuery("pnum", "0")
	perNumsInt, _ := strconv.Atoi(perNums)
	req.PagePerNums = int32(perNumsInt)

	keywords := ctx.DefaultQuery("q", "")
	req.KeyWords = keywords

	brandId := ctx.DefaultQuery("b", "0")
	brandIdInt, _ := strconv.Atoi(brandId)
	req.Brand = int32(brandIdInt)

	//grpc 请求商品的service服务
	r, err := global.GoodsSrvCli.Goods.GoodsList(context.WithValue(context.Background(), "ginContext", ctx), req)
	if err != nil {
		zap.S().Errorw("[GetGoodsList] 查询 【商品列表】失败")
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}

	reMap := map[string]interface{}{
		"total": r.Total,
	}

	goodsList := make([]interface{}, 0)
	for _, value := range r.Data {
		goodsList = append(goodsList, map[string]interface{}{
			"id":          value.Id,
			"name":        value.Name,
			"goods_brief": value.GoodsBrief,
			"desc":        value.GoodsDesc,
			"ship_free":   value.ShipFree,
			"images":      value.Images,
			"desc_images": value.DescImages,
			"front_image": value.GoodsFrontImage,
			"shop_price":  value.ShopPrice,
			"ctegory": map[string]interface{}{
				"id":   value.Category.Id,
				"name": value.Category.Name,
			},
			"brand": map[string]interface{}{
				"id":   value.Brand.Id,
				"name": value.Brand.Name,
				"logo": value.Brand.Logo,
			},
			"is_hot":  value.IsHot,
			"is_new":  value.IsNew,
			"on_sale": value.OnSale,
		})
	}
	reMap["data"] = goodsList

	ctx.JSON(http.StatusOK, reMap)
}

func NewGoods(ctx *gin.Context) {
	//新增商品
	goodsForm := forms.GoodsForm{}
	if err := ctx.ShouldBindJSON(&goodsForm); err != nil {
		api.HandleValidatorError(ctx, err)
		return
	}
	//通过grpc调用商品的service服务
	rsp, err := global.GoodsSrvCli.Goods.CreateGoods(context.Background(), &proto.CreateGoodsInfo{
		Name:            goodsForm.Name,
		GoodsSn:         goodsForm.GoodsSn,
		Stocks:          goodsForm.Stocks,
		MarketPrice:     goodsForm.MarketPrice,
		ShopPrice:       goodsForm.ShopPrice,
		GoodsBrief:      goodsForm.GoodsBrief,
		ShipFree:        *goodsForm.ShipFree,
		Images:          goodsForm.Images,
		DescImages:      goodsForm.DescImages,
		GoodsFrontImage: goodsForm.FrontImage,
		CategoryId:      goodsForm.CategoryId,
		BrandId:         goodsForm.Brand,
	})
	if err != nil {
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}
	ctx.JSON(http.StatusOK, rsp)

}

func GetGoodsDetail(ctx *gin.Context) {
	id := ctx.Param("id")
	idINt, err := strconv.Atoi(id)
	if err != nil {
		ctx.Status(http.StatusNotFound)
		return
	}
	rsp, err := global.GoodsSrvCli.Goods.GetGoodsDetail(context.Background(), &proto.GoodInfoRequest{
		Id: int32(idINt),
	})
	if err != nil {
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}
	rsps := map[string]interface{}{
		"id":          rsp.Id,
		"name":        rsp.Name,
		"goods_brief": rsp.GoodsBrief,
		"desc":        rsp.GoodsDesc,
		"ship_free":   rsp.ShipFree,
		"images":      rsp.Images,
		"desc_images": rsp.DescImages,
		"front_image": rsp.GoodsFrontImage,
		"shop_price":  rsp.ShopPrice,
		"ctegory": map[string]interface{}{
			"id":   rsp.Category.Id,
			"name": rsp.Category.Name,
		},
		"brand": map[string]interface{}{
			"id":   rsp.Brand.Id,
			"name": rsp.Brand.Name,
			"logo": rsp.Brand.Logo,
		},
		"is_hot":  rsp.IsHot,
		"is_new":  rsp.IsNew,
		"on_sale": rsp.OnSale,
	}
	ctx.JSON(http.StatusOK, rsps)
}

func DeleteGoods(ctx *gin.Context) {
	id := ctx.Param("id")
	i, err := strconv.ParseInt(id, 10, 32)
	if err != nil {
		ctx.Status(http.StatusNotFound)
		return
	}
	_, err = global.GoodsSrvCli.Goods.DeleteGoods(context.Background(), &proto.DeleteGoodsInfo{Id: int32(i)})
	if err != nil {
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}

	ctx.Status(http.StatusOK)
}

func Stocks(ctx *gin.Context) {
	id := ctx.Param("id")
	idInt, err := strconv.ParseInt(id, 10, 32)
	if err != nil {
		ctx.Status(http.StatusNotFound)
		return
	}

	resp, err := global.InventorySrvCli.InvDetail(context.Background(), &proto.GoodsInvInfo{
		GoodsId: int32(idInt),
	})
	if err != nil {
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}

	ctx.JSON(http.StatusOK, gin.H{
		"goodsId": resp.GoodsId,
		"num":     resp.Num,
	})
}

// 更新部分状态 GoodsStatusForm字段
func UpdateGoodsStatus(ctx *gin.Context) {
	goodsStatusForm := forms.GoodsStatusForm{}
	if err := ctx.ShouldBindJSON(&goodsStatusForm); err != nil {
		api.HandleValidatorError(ctx, err)
		return
	}

	id := ctx.Param("id")
	i, err := strconv.ParseInt(id, 10, 32)
	if _, err = global.GoodsSrvCli.Goods.UpdateGoodsStatus(context.Background(), &proto.GoodsStatusRequest{
		Id:     int32(i),
		IsHot:  *goodsStatusForm.IsHot,
		IsNew:  *goodsStatusForm.IsNew,
		OnSale: *goodsStatusForm.OnSale,
	}); err != nil {
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}
	ctx.JSON(http.StatusOK, gin.H{
		"msg": "修改成功",
	})
}

// 更新商品
func UpdateGoods(ctx *gin.Context) {
	goodsForm := forms.GoodsForm{}
	if err := ctx.ShouldBindJSON(&goodsForm); err != nil {
		api.HandleValidatorError(ctx, err)
		return
	}

	id := ctx.Param("id")
	i, err := strconv.ParseInt(id, 10, 32)
	if _, err = global.GoodsSrvCli.Goods.UpdateGoods(context.Background(), &proto.CreateGoodsInfo{
		Id:              int32(i),
		Name:            goodsForm.Name,
		GoodsSn:         goodsForm.GoodsSn,
		Stocks:          goodsForm.Stocks,
		MarketPrice:     goodsForm.MarketPrice,
		ShopPrice:       goodsForm.ShopPrice,
		GoodsBrief:      goodsForm.GoodsBrief,
		ShipFree:        *goodsForm.ShipFree,
		Images:          goodsForm.Images,
		DescImages:      goodsForm.DescImages,
		GoodsFrontImage: goodsForm.FrontImage,
		CategoryId:      goodsForm.CategoryId,
		BrandId:         goodsForm.Brand,
	}); err != nil {
		api.HandleGrpcErrorToHttpError(err, ctx)
		return
	}
	ctx.JSON(http.StatusOK, gin.H{
		"msg": "更新成功",
	})
}

File: atopmall_srvs/goods_srv/handler/goods.py
import grpc
from loguru import logger
from peewee import DoesNotExist
from google.protobuf import empty_pb2
from goods_srv.model.models import Goods,Category,Brands
from goods_srv.proto import goods_pb2,goods_pb2_grpc

# 商品服务
class GoodsServicer(goods_pb2_grpc.GoodsServicer):
        
    def convert_model_to_message(self, goods):
        info_rsp = goods_pb2.GoodsInfoResponse()

        info_rsp.id = goods.id
        info_rsp.categoryId = goods.category_id
        info_rsp.name = goods.name
        info_rsp.goodsSn = goods.goods_sn
        info_rsp.clickNum = goods.click_num
        info_rsp.soldNum = goods.sold_num
        info_rsp.favNum = goods.fav_num
        info_rsp.marketPrice = goods.market_price
        info_rsp.shopPrice = goods.shop_price
        info_rsp.goodsBrief = goods.goods_brief
        info_rsp.shipFree = goods.ship_free
        info_rsp.goodsFrontImage = goods.goods_front_image
        info_rsp.isNew = goods.is_new
        info_rsp.descImages.extend(goods.desc_images)
        info_rsp.images.extend(goods.images)      #使用的extend方法而不是append方法，是因为goods.images是一个列表，而info_rsp.images是一个列表，需要将goods.images中的元素添加到info_rsp.images中，合并两个列表
        info_rsp.isHot = goods.is_hot
        info_rsp.onSale = goods.on_sale

        info_rsp.category.id = goods.category.id
        info_rsp.category.name = goods.category.name

        info_rsp.brand.id = goods.brand.id
        info_rsp.brand.name = goods.brand.name
        info_rsp.brand.logo = goods.brand.logo

        return info_rsp

    @logger.catch # 商品列表页
    def GoodsList(self, request:goods_pb2.GoodsFilterRequest, context):
        #商品列表页
        rsp = goods_pb2.GoodsListResponse() #存储查询到的商品列表信息返回给客户端
        
        #过滤条件
        goods = Goods.select() #查询所有商品
        if request.keyWords: #一个简单的搜索
            goods = goods.filter(Goods.name.contains(request.keyWords))  #filter()方法用于过滤查询结果，返回符合条件的商品列表 where()方法用于指定查询条件，返回符合条件的商品列表
        if request.isHot:
            goods = goods.filter(Goods.is_hot == True)
        if request.isNew:
            goods = goods.filter(Goods.is_new == True)
        if request.priceMin:
            goods = goods.filter(Goods.shop_price >= request.priceMin)
        if request.priceMax:
            goods = goods.filter(Goods.shop_price <= request.priceMax)
        if request.brand:
            goods = goods.filter(Goods.brand_id == request.brand)
        #分类级别过滤 通过category查询商品，可能是一级分类，可能是二级分类，可能是三级分类
        if request.topCategory:
            try:
                ids=[]
                category = Category.get(Category.id == request.topCategory)
                level = category.level
                if level == 1: 
                    #SELECT * FROM category WHERE parent_category_id IN ( SELECT category.id FROM category WHERE parent_category_id=130358)
                    categorys = Category.select().where(Category.parent_category_id.in_(Category.select().where(Category.parent_category_id == request.topCategory)))
                    for category in categorys:
                        ids.append(category.id)
                elif level == 2:
                    categorys = Category.select().where(Category.parent_category_id == request.topCategory)
                    for category in categorys:
                        ids.append(category.id)
                elif level == 3:
                    ids.append(request.topCategory)
                goods = goods.filter(Goods.category_id.in_(ids))
            except DoesNotExist:
                logger.error(f"分类id={request.topCategory}不存在")
                rsp.total = 0
                return rsp
        #分页 limit offset
        start = 0
        page = 1
        pagePerNums = 10
        if request.pagePerNums:
            pagePerNums = request.pagePerNums
        if request.pages:
            page = request.pages
            start = (page-1)*pagePerNums

        rsp.total = goods.count()
        goods = goods.limit(pagePerNums).offset(start)
        for good in goods:
            rsp.data.append(self.convert_model_to_message(good))

        return rsp
    
    @logger.catch # 批量获取商品详情
    def BatchGetGoods(self,request:goods_pb2.BatchGoodsIdInfo,context):
        #批量获取商品详情,订单新建的时候可以使用
        rsp = goods_pb2.GoodsListResponse()
        goods = Goods.select().where(Goods.id.in_(list(request.id))) #根据商品id列表查询所有商品
        rsp.total = goods.count()
        for good in goods:
            rsp.data.append(self.convert_model_to_message(good))
        return rsp
    
    @logger.catch # 删除商品
    def DeleteGoods(self,request:goods_pb2.DeleteGoodsInfo,context):
        #删除商品
        # rows = Goods.delete().where(Goods.id==request.id)
        # if rows == 0:
        #     logger.error(f"商品id={request.id}不存在")
        #     return empty_pb2.Empty()
        try:
            goods =  Goods.get(Goods.id == request.id)
            goods.delete_instance() #删除商品
            return empty_pb2.Empty()
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details(f"商品id={request.id}不存在")
            return empty_pb2.Empty()
        except Exception as e:
            context.set_code(grpc.StatusCode.INTERNAL) #严重的内部错误
            context.set_details(str(e))
            return empty_pb2.Empty()
    
    @logger.catch # 获取商品详情
    def GetGoodsDetail(self,request:goods_pb2.GoodInfoRequest,context):
        #获取商品的详情
        try:
            goods =  Goods.get(Goods.id == request.id)
            #每次查询商品详情，商品的点击量+1
            goods.click_num += 1
            goods.save() #保存商品信息
            rsp = self.convert_model_to_message(goods)
            return rsp
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details(f"商品id={request.id}不存在")
            return goods_pb2.GoodsInfoResponse()
        
    @logger.catch # 创建商品
    def CreateGoods(self,request:goods_pb2.CreateGoodsInfo,context):
        #创建商品
        #先处理外键 商品分类和品牌分类
        try:
            category = Category.get(Category.id == request.categoryId)
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("商品分类不存在")
            return goods_pb2.GoodsInfoResponse()
        try:
            brand = Brands.get(Brands.id == request.brandId)
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("品牌不存在")
            return goods_pb2.GoodsInfoResponse()
        
        goods = Goods() #创建商品实例 将grpc消息转换为数据库模型存储到数据库
        goods.brand = brand
        goods.category = category
        goods.name = request.name
        goods.goods_sn = request.goodsSn
        goods.market_price = request.marketPrice
        goods.shop_price = request.shopPrice
        goods.goods_brief = request.goodsBrief
        goods.ship_free = request.shipFree
        goods.images = list(request.images)
        goods.desc_images = list(request.descImages)
        goods.goods_front_image = request.goodsFrontImage
        goods.is_new = request.isNew
        goods.is_hot = request.isHot
        goods.on_sale = request.onSale

        goods.save()

        #TODO 此处完善库存的设置 - 分布式事务
        return self.convert_model_to_message(goods)
    
    @logger.catch # 更新商品
    def UpdateGoods(self,request:goods_pb2.CreateGoodsInfo,context):
        #更新商品
        #先处理外键 商品分类和品牌分类和查询商品是否存在
        try:
            category = Category.get(Category.id == request.categoryId)
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("商品分类不存在")
            return goods_pb2.GoodsInfoResponse()
       
        try:
            brand = Brands.get(Brands.id == request.brandId)
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("品牌不存在")
            return goods_pb2.GoodsInfoResponse()

        try:
            goods = Goods.get(Goods.id == request.id)
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("商品不存在")
            return goods_pb2.GoodsInfoResponse()

        goods.brand = brand
        goods.category = category
        goods.name = request.name
        goods.goods_sn = request.goodsSn
        goods.market_price = request.marketPrice
        goods.shop_price = request.shopPrice
        goods.goods_brief = request.goodsBrief
        goods.ship_free = request.shipFree
        goods.images = list(request.images)
        goods.desc_images = list(request.descImages)
        goods.goods_front_image = request.goodsFrontImage
        goods.is_new = request.isNew
        goods.is_hot = request.isHot
        goods.on_sale = request.onSale

        goods.save()

        #TODO 此处完善库存的设置 - 分布式事务
        return self.convert_model_to_message(goods)

    @logger.catch 
    def UpdateGoodsStatus(self, request: goods_pb2.GoodsStatusRequest, context):
        # 更新商品状态（仅更新 is_new/is_hot/on_sale）
        try:
            goods = Goods.get(Goods.id == request.id)
        except DoesNotExist:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            context.set_details("商品不存在")
            return empty_pb2.Empty()

        goods.is_new = request.isNew
        goods.is_hot = request.isHot
        goods.on_sale = request.onSale
        goods.save()

        return empty_pb2.Empty()